nixos/lib/systemd-unit-options.nix

```diff
@@ -231,6 +231,79 @@ in rec {
   };
 
 
+  systemdCommandExpansion = types.attrsOf types.str // {
+    name = "systemdCommandExpansion";
+    description = "systemd-expanded string";
+    check = x: isAttrs x && elem (attrNames x) [ ["env"] ["substitute"] ];
+    merge = mergeOneOption;
+  };
+
+  serviceCommandType = types.submodule ({ config, options, ... }: {
+    options = {
+      argv0 = mkOption {
+        default = null;
+        type = with types; nullOr (oneOf [ str systemdCommandExpansion ]);
+        description = ''
+          Override argv[0] (the executable name given to the command).
+        '';
+      };
+
+      mayFail = mkOption {
+        default = false;
+        type = types.bool;
+        description = ''
+          Do not consider it a failure when this command returns a non-zero
+          exit status.
+        '';
+      };
+
+      unrestricted = mkOption {
+        default = null;
+        type = with types; nullOr (enum [ "all" "credentials" ]);
+        description = ''
+          If "all": run this command with full privileges, regardless of configured
+          privilege restrictions (equivalent to the system "+" command prefix).
+
+          If "credentials": do not apply user and group settings when running the
+          command (equivalent to the systemd "!" command prefix).
+        '';
+      };
+
+      exe = mkOption {
+        type = with types; nullOr str;
+        description = ''
+          Command to run. Mutually exclusive with `script`.
+        '';
+      };
+
+      script = mkOption {
+        default = null;
+        type = with types; nullOr lines;
+        description = ''
+          Script to run. Mutually exclusive with `exe`.
+        '';
+      };
+
+      args = mkOption {
+        default = [ ];
+        type = with types; listOf (oneOf [ str systemdCommandExpansion int float ]);
+        description = ''
+          Arguments passed to the command. Arguments will not be expanded
+          by systemd unless requested.
+
+          To expand an environment variable `X` use `{ env = "X"; }`.
+
+          To apply all systemd substitutions to a string use `{ substitute = "..."; }`.
+        '';
+      };
+    };
+
+    config = {
+      exe = mkIf (config.script != null)
+        (mkDerivedConfig (options.script) builtins.toFile);
```

Comment:
You'll need a `writeScriptBin` here, both to make it executable and to allow store path references.